libp2pdht/private/eth/p2p/discoveryv5/protocol.nim

```diff
@@ -135,7 +138,7 @@ type
     talkProtocols*: Table[seq[byte], TalkProtocol] # TODO: Table is a bit of
     # overkill here, use sequence
     rng*: ref BrHmacDrbgContext
-    providers: Table[NodeId, seq[SignedPeerRecord]]
+    providers: SQLiteDatastore
```

Comment:
This should definitely be `Datastore` and it would be initialized with a `TieredDatastore `, which should in turn contain an `LRUDatastore` and the `SQLiteDatastore`. Afaik, we're missing the `LRUDatastore`, but it should be relatively straight forward to implement one based on the implementation we're using in codex proper. It might even just be a wrapper around it.

Reply:
I would rather make a copy and adapt it to Nim-Datastore API. Eventually, it will be great to move all but Network stores from Codex to Nim-Datastore and drop Codex own blockstore API entirely.

Reply:
Good catch re: providers, I meant to change it earlier but forgot.

`LRUDatastore` is easy to implement by combining with e.g. status-im/lrucache.nim, but there are some things to consider, which is why I held off doing that. I wrote a lot in the PR description above, so will try to summarize:

(1)

Since the Datastore API is not generic, the data parameter will need to be `seq[byte]`. It's not a huge problem, but if the caller is concerned about how much memory the full cache will use, then they will need to have some idea about the length of the data blobs in order to specify an appropriate `size`. Seems clunky, and I thought maybe it's best to wait on `LRUDatastore` until we make the nim-datastore API generic using vtables approach or however we do it.

(2)

We are dealing with a one (NodeId) to many (SPR) relationship. So how are we going to represent that in a key-value store? In a `TieredDatastore` setup, the way we choose to do it will impact all of the tiers.

(a)

The way I did it in this PR is to build each key by combining the NodeId (Cid) with a fast hash of the SPR:
```
/[cid]/[spr-hash]
```
That converts the one-to-many relationship to one-to-one. But that means that SPR lookup becomes a `query` method call for `/[cid]/*`.

In a `TieredDatastore` the only source of truth for `query` is the bottom tier; for us that would mean skipping the LRU cache and querying SQLite. I didn't come up with the idea of limiting queries to the bottom tier, it's in the [doc comments](https://github.com/datastore/datastore/blob/master/datastore/core/basic.py#L1035) for the original Python datastore. It does make sense, though, because it's hard to understand what it would look like to orchestrate a query through 2+ tiers.

Thinking about that problem is what led me to suggest SQLite's own page cache (configured to our preferred size) might do a good enough job such that we don't need to implement a caching layer in Nim.

(b)

Another way to deal with one-to-many is to store the values (SPRs) in a seq and serialize them as one big blob. That can work, but there are tradeoffs.

`LRUDatastore` might be too coarse: some NodeIds may have a lot of SPRs, others just a few, but cache eviction only considers the LRU aspect for NodeId.

If we switch to a two-level cache as proposed in #40, then we have to depart from `TieredDatastore`. That would be okay, but see #43 or my long comment in #40 about the decision that forces on us.

Reply:
In https://github.com/status-im/nim-datastore/pull/27 I made changes to support changing `providers: SQLiteDatastore` to `providers: Datastore`, and I've already updated this PR to make use of those changes.

A PR introducing `LRUDatastore` will be submitted shortly.